File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/SkipDateUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.model.SkipException
import com.nogeass.passalarm.domain.model.SkipReason
import com.nogeass.passalarm.domain.repository.SkipExceptionRepository
import javax.inject.Inject

class SkipDateUseCase @Inject constructor(
    private val skipRepository: SkipExceptionRepository,
    private val reschedule: RescheduleNextNUseCase
) {
    suspend fun execute(planId: Long, date: String, reason: SkipReason = SkipReason.MANUAL) {
        skipRepository.save(SkipException(planId = planId, date = date, reason = reason))
        reschedule()
    }

    suspend fun unskip(planId: Long, date: String) {
        skipRepository.deleteByPlanAndDate(planId, date)
        reschedule()
    }
}

File: apps/ios/PassAlarm/Domain/UseCase/SkipDateUseCase.swift
import Foundation

final class SkipDateUseCase: Sendable {
    private let skipRepository: SkipExceptionRepositoryProtocol
    private let reschedule: RescheduleNextNUseCase

    init(skipRepository: SkipExceptionRepositoryProtocol,
         reschedule: RescheduleNextNUseCase) {
        self.skipRepository = skipRepository
        self.reschedule = reschedule
    }

    func execute(planId: UUID, date: String, reason: SkipException.SkipReason = .manual) async throws {
        let skip = SkipException(
            id: UUID(),
            planId: planId,
            date: date,
            reason: reason,
            createdAt: Date()
        )
        try await skipRepository.save(skip)
        try await reschedule.execute()
    }

    func unskip(planId: UUID, date: String) async throws {
        try await skipRepository.deleteByPlanAndDate(planId: planId, date: date)
        try await reschedule.execute()
    }
}
